code/espurna/kacurtainswitch.ino

```diff
@@ -0,0 +1,162 @@
+#define KA_CURTAIN_PORT  KA_CURTAIN_HW_PORT
+
+// -----------------------------------------------------------------------------
+// Private
+// -----------------------------------------------------------------------------
+
+void _KACurtainReceiveUART() {
+    static unsigned char ndx = 0;
+    while (KA_CURTAIN_PORT.available() > 0 && _KACurtainNewData == false) {
+        char rc = KA_CURTAIN_PORT.read();
+        if (rc != KA_CURTAIN_TERMINATION) {
+            _KACurtainBuffer[ndx] = rc;
+            if (ndx < KA_CURTAIN_BUFFER_SIZE - 1) ndx++;
+        } else {
+            _KACurtainBuffer[ndx] = '\0';
+            _KACurtainNewData = true;
+            ndx = 0;
+        }
+    }
+}
+
+void _kacurtainSendMQTT() {
+    if (_KACurtainNewData == true && MQTT_SUPPORT) {
+        DEBUG_MSG_P(PSTR("[KACURTAIN] Send data over MQTT: %s\n"), _KACurtainBuffer);
+        mqttSend(MQTT_TOPIC_CURTAININ, _KACurtainBuffer);
+        _KACurtainNewData = false;
+        if (String(_KACurtainBuffer).indexOf("enterESPTOUCH") > 0 ) {
+          wifiStartAP();
+        } else if (String(_KACurtainBuffer).indexOf("exitESPTOUCH") > 0 ) {
+          deferredReset(100, CUSTOM_RESET_HARDWARE);
+        }
+    }
+}
+
+void _KACurtainActionSelect(const char * message) {
+  if (String(message) == "pause") {
+    _KACurtainPause(message);
+  } else {
+    _KACurtainSetclose(message);
+  }
+}
+
+void _KACurtainPause(const char * message) {
+  // Tell N76E003AT20 to stop moving and report current position
+  DEBUG_MSG_P(PSTR("[KACURTAIN] message: %s\n"), message);
+  char tx_buffer[80];
+  snprintf_P(tx_buffer, sizeof(tx_buffer), PSTR("AT+UPDATE=\"sequence\":\"%d%03d\",\"switch\":\"%s\""),
+   now(), millis()%1000, message);
+  _KACurtainSend(tx_buffer);
+}
+
+void _KACurtainSetclose(const char * message) {
+  // Tell N76E003AT20 to go to position X (based on X N76E003AT20 decides to go up or down)
+  DEBUG_MSG_P(PSTR("[KACURTAIN] payload: %s\n"), message);
+  char tx_buffer[80];
+  // %d   = long / uint8_t
+  // %03d = long / uint8_t - last 3 digits
+  // %s   = char
+  snprintf_P(tx_buffer, sizeof(tx_buffer), PSTR("AT+UPDATE=\"sequence\":\"%d%03d\",\"switch\":\"%s\",\"setclose\":%s"),
+   now(), millis()%1000, "off", message);
+  _KACurtainSend(tx_buffer);
+}
+
+void _KACurtainSendOk() {
+  // Confirm N76E003AT20 message received and stop repeating
+  if (_KACurtainNewData == true) {
+    DEBUG_MSG_P(PSTR("[KACURTAIN] send ok"));
+    KA_CURTAIN_PORT.print("AT+SEND=ok");
+    KA_CURTAIN_PORT.write(0x1B);
+    KA_CURTAIN_PORT.flush();
+  }
+}
+
+void _KACurtainSend(const char * tx_buffer) {
+  KA_CURTAIN_PORT.print(tx_buffer);
+  KA_CURTAIN_PORT.write(0x1B);
+  KA_CURTAIN_PORT.flush();
+}
+
+void _KACurtainCallback(unsigned int type, const char * topic, const char * payload) {
+    if (type == MQTT_CONNECT_EVENT) {
+        mqttSubscribe(MQTT_TOPIC_CURTAINOUT);
+        DEBUG_MSG_P(PSTR("[KACURTAIN] Subscribe: %s\n"), MQTT_TOPIC_CURTAINOUT);

```

Comment:
mqttSubscribe does log already

Reply:
Do you mean these debug line is not needed?

Reply:
changed